### DISCUSSION TITLE
Shortcuts for custom functions at event / note level

### DISCUSSION THREAD

**Discussion by nilhartman** [tidalcycles/Tidal#1167] [Ideas]
Hi!
Hopefully I'll succeed in keeping it as concise and clear as possible. If doable, it'd make TC an even terser langage than what it is already, and would open a world of new possibilities.

Anyway, here we go :

"?" is an alias / shortcut for "degradeBy 0.5", right?  

 "?" can be used inside a pattern, as in :

do
setcps (160/60/4)
d1 $ s "[bd ~ ~ bd ~ ~ bd? ~, ~ [rs rs? rs rs], cp:1*16?]"

 ... while "degradeBy 0.5" cannot, it has to be written prior to a pattern, and then all the events of the pattern can be potentially altered. Still correct?

Same goes for customs functions (those you've defined using "let"), you have to write them before a pattern to make them work.  

Now, would it possible in any way to create our own aliases for such custom functions (shortcuts that could be used similarly as "?" )?   

Something like :  

do
setcps (160/60/4)
d1 $ s "[bd ~ ~ bd€ ~ ~ bd? ~, ~ [rs rs£ rs rs]•, cp:1*16¥]"

Where, for example, € would equal "degradeBy 0.125", £ = "degradeBy 0.2345", • = "$ every 8 (rev)" and ¥ = "speed "[1 1.5 0.75 -1.25]/4""... or whatever, as long as you've defined those custom function / aliases.

So that all sorts of pattern modifications could happen on a step / note / event basis.

Thanks for reading !

Best, 
Nil.

**Comment by yaxu** [comment]
Hi @nilhartman, thanks for the suggestion, a nice idea.
It's a bit similar to `ur`, which allows you to name functions and then apply them to events in the sequence. In the case of `ur`, events are themselves patterns but perhaps a similar approach could be taken here..
